### ISSUE TITLE
Help: "Make" fails on Ubuntu 17.10

### ISSUE THREAD

**Issue by HirIMogul** [intel/QAT_Engine#55]
Hey Steve,
I tried to compile the Intel QAT 1.7 driver on Ubuntu 17.10 following the README file in the [archive](https://01.org/sites/default/files/downloads/intelr-quickassist-technology/qat1.7.upstream.l.1.0.3-42.tar.gz), as well as the provided [PDF](https://01.org/sites/default/files/downloads/intelr-quickassist-technology/336212qatswgsgrev002review.pdf). Unfortunately the build failed with an error I was not able to interpret. This is most likely because I just started using Linux, but I try hard to improve and I would be very happy, if you could assist me doing this.

I installed all dependencies according to the README file.
I run ./[configure](https://github.com/intel/QAT_Engine/files/1693077/configure.txt)
I run [make](https://github.com/intel/QAT_Engine/files/1693078/make.txt)

Thank you very much for your help.

**Comment by stevelinsell** [comment]
Hi @HirlMogul,

We have run recently on Ubuntu with the LTS release 16.04.03 (Xenial Xerus) without issue but we do not run with the very latest 17.10.01 Ubuntu release. Saying that I would not expect there to be any issues with the latest.
I'm not an expert at getting the QAT Driver compiling and running (I work and support the application layer above this) but if you can provide the output of the ./configure and the make command I can take a look.
As a first step have you ensured you have installed all the dependencies described in section 3.2 of the PDF file you mention. Please note that on Ubuntu you won't use `yum` to install the packages but `apt-get install`. The libraries you want to install have slightly different names under Ubuntu too, the ones you want are:
pciutils
libudev-dev
libboost-dev
build-essential
zlib1g-dev
libssl-dev
 
Kind Regards,

Steve.

**Comment by stevelinsell** [comment]
Hi @HirIMogul,

I forgot to mention you may also need to install the following package:
linux-headers-generic
but try without first.

Kind Regards,

Steve.

**Comment by HirIMogul** [comment]
Hey Steve. I provided the requested files (configure / make) in my first post as links in text. Just click on the links to download the text files.

I checked for the dependencies. They are all fine.
```
sudo apt-get install pciutils libudev1 libudev-dev libboost-dev build-essential zlib1g-dev libssl-dev linux-headers-generic
Reading package lists... Done
Building dependency tree
Reading state information... Done
build-essential is already the newest version (12.4ubuntu1).
libboost-dev is already the newest version (1.62.0.1).
pciutils is already the newest version (1:3.5.2-1ubuntu1).
zlib1g-dev is already the newest version (1:1.2.11.dfsg-0ubuntu2).
libudev-dev is already the newest version (234-2ubuntu12.1).
libudev1 is already the newest version (234-2ubuntu12.1).
linux-headers-generic is already the newest version (4.13.0.32.34).
linux-headers-generic set to manually installed.
libssl-dev is already the newest version (1.1.0g-2.1+ubuntu17.10.1+deb.sury.org+1).
0 upgraded, 0 newly installed, 0 to remove and 20 not upgraded.
```

Thanks for helping me out with it.

**Comment by stevelinsell** [comment]
Hi @HirIMogul,

My apologies I missed the links in your original post.
I've recreated your issue locally and investigated.
It appears the kernel headers for crypto changed between 4.12.0 and 4.13.0.
Specifically the akcipher_alg struct, within akcipher.h changed.
The function pointer for max_size changed signature to return an unsigned int rather than an int.
The same is true for the kpp_alg struct within kpp.h where max_size changed in the same way.
Also the set_secret function within kpp_alg changed to constify the void * parameter. This change was slightly earlier between the 4.11.0 and 4.12.0 kernel.

I will raise a defect on the QAT Driver team to address these discrepencies going forward.
In the meantime you have two choices:

1) Use an earlier kernel. I would recommend using 4.9.80 as at the time of writing that is the latest long term support kernel branch.
2) Make some changes yourself to the QAT Driver to make the functions being assigned to the function pointers have the correct types (and change the functions themselves to use the correct types internally). I haven't tried this myself yet so I can't comment on how easy that will be but hopefully it is quite straightforward.

Hope this info is helpful,

Steve.

**Comment by HirIMogul** [comment]
Hey Steve. I'll wait until a new version is released. Downgrading the kernel would result in my network interface not being recognized. I hope that with the LTS release of Ubuntu 18.04 the driver is going to run fine.

Out of curiosity. Is the QAT driver pre-installed in Intels clear Linux distribution? If not, is it supported to install it the same way as in Ubuntu?

**Comment by HirIMogul** [comment]
The latest long-kernel has been released a week ago. Is kernel 4.14.20 supported by the current release?

**Comment by stevelinsell** [comment]
Hi @HirIMogul,

Just a quick update. I checked and the QuickAssist Driver master branch has already been updated to fix the issues you encountered and also an issue with the compilation of the USDM Memory Driver on newer kernels. Unfortunately I do not have a date for you when these fixes will be visible in a release on the 01.org website.  I checked with the QuickAssist Driver development team and they have not tested the master branch QuickAssist Driver against kernel 4.14.20 yet, obviously the current version on 01.org will exhibit at least the same issues you saw with Ubuntu 17.10 and a 4.13.X kernel.

Kind Regards,

Steve.
